examples/scripts/spme_max_energy.py

```diff
@@ -0,0 +1,175 @@
+    negative_mass_density = (
+        ps["Negative electrode active material volume fraction"]
+        * ps["Negative electrode density [kg.m-3]"]
+    )
+    +(ps["Negative electrode porosity"] * electrolyte_density)
+
+    # Areal mass densities [kg.m-2]
+    positive_mass = ps["Positive electrode thickness [m]"] * positive_mass_density
+    negative_mass = ps["Negative electrode thickness [m]"] * negative_mass_density
+    separator_mass = (
+        ps["Separator thickness [m]"]
+        * ps["Separator porosity"]
+        * ps["Separator density [kg.m-3]"]
+    )
+    positive_current_collector_mass = (
+        ps["Positive current collector thickness [m]"]
+        * ps["Positive current collector density [kg.m-3]"]
+    )
+    negative_current_collector_mass = (
+        ps["Negative current collector thickness [m]"]
+        * ps["Negative current collector density [kg.m-3]"]
+    )
+
+    # Cross-sectional area [m2]
+    cross_sectional_area = ps["Electrode height [m]"] * ps["Electrode width [m]"]
+
+    return cross_sectional_area * (
+        positive_mass
+        + separator_mass
+        + negative_mass
+        + positive_current_collector_mass
+        + negative_current_collector_mass
+    )
+
+
+# Define test protocol
+experiment = pybamm.Experiment(
+    ["Discharge at 1C until 2.5 V (5 seconds period)"],
+)
+init_soc = 1  # start from full charge
+signal = ["Voltage [V]", "Current [A]"]
+
+# Generate problem
+problem = pybop.DesignProblem(
+    model, parameters, experiment, signal=signal, init_soc=init_soc
+)
+
+# Update the C-rate and the example dataset
+nominal_capacity(problem.x0, model)
+sol = problem.evaluate(problem.x0)
+problem._time_data = sol[:, -1]
+problem._target = sol[:, 0:-1]
+
+
+# Define cost function as a subclass
+class GravimetricEnergyDensity(pybop.BaseCost):
+    """
+    Defines the (negative*) gravimetric energy density corresponding to a
+    normalised 1C discharge from upper to lower voltage limits.
+    *The energy density is maximised by minimising the negative energy density.
+    """
+
+    def __init__(self, problem):
+        super().__init__(problem)
+
+    def _evaluate(self, x, grad=None):
+        """
+        Compute the cost
+        """
+        with warnings.catch_warnings(record=True) as w:
+            # Update the C-rate and run the simulation
+            nominal_capacity(x, self.problem._model)
+            sol = self.problem.evaluate(x)
+
+            if any(w) and issubclass(w[-1].category, UserWarning):
+                # Catch infeasible parameter combinations e.g. E_Li > Q_p
+                return np.inf
+
+            else:
+                voltage = sol[:, 0]
+                current = sol[:, 1]
+                gravimetric_energy_density_Ah = np.trapz(voltage * current) / (
+                    3600 * cell_mass(self.problem._model._parameter_set)
+                )
+                # Take negative in order to maximise energy density
```

Comment:
Might be more clear to users as:

"Returning the negative value as optimiser minimises this function, results in maximising the energy density"